<!--
%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{R/qtl2 user guide}
-->

# R/qtl2 user guide

[R/qtl2](http://kbroman.org/qtl2) (aka qtl2) is a reimplementation of the QTL analysis software
[R/qtl](http://www.rqtl.org), to better handle high-dimensional data
and complex cross designs.

## Installation

R/qtl2 is early in development and so is not yet available on
[CRAN](http://cran.r-project.org).

You can install R/qtl2 from its
[GitHub repository](http://github.com/kbroman/qtl2). You first need to
install the [devtools](https://github.com/hadley/devtools) package.

```{r install_devtools, eval=FALSE}
install.packages("devtools")
```

Then install R/qtl2 using `devtools::install_github()`.

```{r install_qtl2, eval=FALSE}
library(devtools)
install_github("kbroman/qtl2")
```

## Data file format

The input data file formats for [R/qtl](http://www.rqtl.org) cannot
handle complex crosses, and so for R/qtl2, we have defined a new
format for the data files. We'll describe it here briefly; for
details, see the separate
[vignette on the input file format](input_files.html).

QTL mapping data consists of a set of tables of data: marker
genotypes, phenotypes, marker maps, etc. In the new format, these
different tables are in separate comma-delimited (CSV) files. In each
file, the first column is a set of IDs for the rows, and the first row
is a set of IDs for the columns. For example, the phenotype data file
will have individual IDs in the first column and phenotype names in
the first row.

A few important changes in the tabular data:

- We will use not just the genetic marker map, but also a physical map
  (if available).
- Previously, phenotypes and covariates were combined. In the new
  format, we separate numeric phenotypes from the often
  non-numeric covariates.
- We define a table of &ldquo;phenotype covariates.&rdquo; These are
  [metadata](http://en.wikipedia.org/wiki/Metadata) describing the
  phenotypes. For example, in the case of a phenotype measured over
  time, one column in the phenotype covariate data frame could be the
  time of measurement. For gene expression data, we would have columns
  representing chromosome and physical position of genes, as well as
  gene IDs.

In additional to the set of CSV files with the primary data, we need a
separate &ldquo;control&rdquo; file with various control parameters
(or metadata), including the names of all of the other data files and
the genotype codes used in the genotype data file. The control file is
in a specific [YAML](http://www.yaml.org) format. YAML is a
human-readable text file for representing relatively complex
data. (It's much like [JSON](http://json.org), but _much_ more
readable.)

A big advantage of this control file scheme is that it greatly
simplifies the function for reading in the data. That function,
`read_cross2()`, has a _single_ argument: the name (with path) of the
control file. So you can read in data like this:

```{r read_cross2, eval=FALSE}
library(qtl2)
grav2 <- read_cross2("~/my_data/grav2.yaml")
```

The large number of files is a bit cumbersome, so we've made it
possible to use a
[zip file](http://en.wikipedia.org/wiki/Zip_(file_format)) containing
all of the data files, and to read that zip file directly. There's even a
function for creating the zip file:

```{r create_zip, eval=FALSE}
zip_datafiles("~/my_data/grav2.yaml")
```

The `zip_datafiles()` function will read the control file to identify
all of the relevant data files and then zip them up into a file with
the same name and location, but with the extension `.zip` rather than
`.yaml`.

To read the data back in, we use the same `read_cross2()` function,
providing the name (and path) of the zip file rather than the control
file.

```{r read_zip, eval=FALSE}
grav2 <- read_cross2("~/my_data/grav2.zip")
```

This can even be done with remote files.

```{r read_remote_zip, eval=FALSE}
grav2 <- read_cross2("http://kbroman.org/qtl2/assets/sampledata/grav2/grav2.zip")
```

Of course, the other advantage of the zip file is that it is
_compressed_ and so smaller than the combined set of CSV files.

The control file may be confusing for some users. To assist in its
construction, there's a function `write_control_file()` that takes the
large set of control parameters as input and then writes the YAML
control file in the appropriate format.

## Sample data sets

Sample data files are available at
<http://kbroman.org/qtl2/pages/sampledata.html>. Zipped versions of
these datasets are included with the package and can be loaded into R
using the `read_cross2()` function.

In the [package source](https://github.com/kbroman/qtl2), these are
located in
[`qtl2/inst/extdata`](https://github.com/kbroman/qtl2/tree/master/inst/extdata).
In the installed version of the package, they are in `qtl2/extdata`,
within whatever directory your R packages were installed. The base R
function `system.file()` can be used to construct the path to these
files.

For example, one of the sample data sets concerns a gravitropism
phenotype in a set of Arabidopsis recombinant inbred lines (RIL), from
[Moore et al. (2013) Genetics 195:1077-1086](http://www.genetics.org/content/195/3/1077.abstract).
The data are in `qtl2/extdata/grav2.zip`, which can be loaded as
follows:

```{r load_grav2, message=FALSE}
library(qtl2)
grav2 <- read_cross2( system.file("extdata", "grav2.zip", package="qtl2") )
```



<!-- the following to make it look nicer -->
<link href="http://kbroman.org/qtlcharts/assets/vignettes/vignette.css" rel="stylesheet"></link>
