<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>Differences between R/qtl and R/qtl2</title>
    <meta name="description" content="Major differences between R/qtl and R/qtl2">
    <meta name="author" content="Karl Broman">

    <!-- Enable responsive viewport -->
    <meta name="viewport" content="width=device-width, initial-scale=1.0">

    <!-- Le HTML5 shim, for IE6-8 support of HTML elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <!-- Le styles -->
    <link href="http://kbroman.org/qtl2/assets/themes/twitter/bootstrap/css/bootstrap.2.2.2.min.css" rel="stylesheet">
    <link href="http://kbroman.org/qtl2/assets/themes/twitter/css/style.css?body=1" rel="stylesheet" type="text/css" media="all">
    <link href="http://kbroman.org/qtl2/assets/themes/twitter/css/kbroman.css" rel="stylesheet" type="text/css" media="all">

    <!-- Le fav and touch icons -->

    <!-- atom & rss feed -->
    <link href="http://kbroman.org/qtl2nil" type="application/atom+xml" rel="alternate" title="Sitewide ATOM Feed">
    <link href="http://kbroman.org/qtl2nil" type="application/rss+xml" rel="alternate" title="Sitewide RSS Feed">

  </head>

  <body>
    <div class="navbar">
      <div class="navbar-inner">
        <div class="container-narrow">
          <a class="brand" href="http://kbroman.org/qtl2">R/qtl2</a>

          <ul class="nav">
              <li><a href="http://kbroman.org/qtl2/pages/docs.html">documentation</a></li>
              <li><a href="http://kbroman.org/qtl2/assets/vignettes/user_guide.html">user guide</a></li>
              <li><a href="http://kbroman.org/qtl2/assets/vignettes/input_files.html">input files</a></li>
              <li><a href="http://kbroman.org/qtl2/pages/sampledata.html">sample data</a></li>
          </ul>

        </div>
      </div>
    </div>

    <div class="container-narrow">

      <div class="content">


<div class="page-header">
  <h2>Differences between R/qtl and R/qtl2</h2>
</div>

<div class="row-fluid">
  <div class="span12">
<div><div> <!-- useless divs to match 2 extra </div> at the end -->






<p><a href="http://kbroman.org/qtl2">R/qtl2</a> is a reimplementation of the QTL analysis software <a href="https://rqtl.org">R/qtl</a>, to better handle high-dimensional data and complex cross designs.</p>
<p>There have been a number of big changes. We summarize some of them here in order to assist long-time R/qtl users in their transition to R/qtl2.</p>
<div id="new-data-file-formats" class="section level2">
<h2>New data file formats</h2>
<p>The input file format has been completely changed. R/qtl2 allows just a single format, with data split across multiple CSV files, and with a control file (in <a href="http://www.yaml.org">YAML</a> or <a href="http://www.json.org">JSON</a> format).</p>
<p>For more information about the new file format, see the <a href="http://kbroman.org/qtl2/assets/vignettes/input_files.html">related vignette</a>, as well as <a href="http://kbroman.org/qtl2/pages/sampledata.html">sample input data files</a> and the <a href="https://github.com/rqtl/qtl2data">qtl2data repository</a>.</p>
<p>R/qtl cross objects can be converted to the new format with the <code>qtl2geno::convert2cross2()</code> function, and so one can continue to use the R/qtl function <code>read.cross()</code> to read in data and then convert them to the new format:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(qtl)
mycross &lt;-<span class="st"> </span><span class="kw">read.cross</span>(<span class="st">&quot;csv&quot;</span>, <span class="dt">file=</span><span class="st">&quot;mydata.csv&quot;</span>)
<span class="kw">library</span>(qtl2geno)
mycross &lt;-<span class="st"> </span><span class="kw">convert2cross2</span>(mycross)</code></pre></div>
</div>
<div id="new-data-structures" class="section level2">
<h2>New data structures</h2>
<p>The data structures used by R/qtl2 are completely different than those used by R/qtl. The new data structures are no simpler than before, but they tend to be a bit “flatter” (that is, less deeply nested). Most everything is a “list”, and we’re using fewer “attributes” and instead including such things as components at the top level.</p>
</div>
<div id="split-into-multiple-packages" class="section level2">
<h2>Split into multiple packages</h2>
<p>R/qtl2 is not a single package as R/qtl is. Rather it’s split into multiple packages:</p>
<ul>
<li><a href="https://github.com/rqtl/qtl2geno">qtl2geno</a>, for calculating genotype probabilities, imputations, and genetic maps</li>
<li><a href="https://github.com/rqtl/qtl2scan">qtl2scan</a>, for QTL genome scans and related calculations</li>
<li><a href="https://github.com/rqtl/qtl2plot">qtl2plot</a>, for data visualization</li>
<li><a href="https://github.com/rqtl/qtl2convert">qtl2convert</a>, for converting data among the R/qtl2, R/qtl, and <a href="https://www.bioconductor.org/packages/release/bioc/html/DOQTL.html">DOQTL</a> formats</li>
</ul>
<p>We have in mind that, for high-dimensional data, the QTL genotype probability calculations with qtl2geno will be performed once and saved, and that the genome scans with qtl2scan will be performed in “batch” (e.g., on a cluster) and also saved, and that interactive analyses will mostly be in the data visualizations with qtl2plot.</p>
<p>It can be confusing to remember which function is in which package. For this reason, we created an additional, largely empty package <a href="https://github.com/rqtl/qtl2">qtl2</a>. If you load qtl2 with <code>library(qtl2)</code>, the three main packages, <a href="https://github.com/rqtl/qtl2geno">qtl2geno</a>, <a href="https://github.com/rqtl/qtl2scan">qtl2scan</a>, and <a href="https://github.com/rqtl/qtl2plot">qtl2plot</a>, will all be loaded.</p>
</div>
<div id="new-functions-names" class="section level2">
<h2>New functions names</h2>
<p>The names of all of the main functions have changed, mostly with periods replaced by underscores. For example:</p>
<ul>
<li><code>read.cross()</code> → <code>read_cross2()</code></li>
<li><code>calc.genoprob()</code> → <code>calc_genoprob()</code></li>
<li><code>sim.geno()</code> → <code>sim_geno()</code></li>
<li><code>argmax.geno()</code> → <code>viterbi()</code></li>
<li><code>scanone()</code> → <code>scan1()</code></li>
<li><code>nmar()</code> → <code>n_mar()</code></li>
<li><code>phenames()</code> → <code>pheno_names()</code></li>
<li><code>markernames()</code> → <code>marker_names()</code></li>
</ul>
</div>
<div id="treatment-of-intermediate-calculations" class="section level2">
<h2>Treatment of intermediate calculations</h2>
<p>We’re no longer storing intermediate calculations as part of the cross object. For example, <code>calc_genoprob()</code>, to calculate QTL genotype probabilities given the observed marker data, returns a list with the probabilities. <code>scan1()</code>, to perform a genome scan, takes these probabilities plus a phenotype matrix.</p>
</div>
<div id="use-of-individual-identifiers-for-aligning-data" class="section level2">
<h2>Use of individual identifiers for aligning data</h2>
<p>Individual identifiers are now used to ensure the alignment of individuals, for example between the QTL genotype probabilties and the phenotype data.</p>
<p>For example, in <code>scan1()</code>, to perform a genome scan, it’s necessary that the phenotype data carry the corresponding individual IDs as row names.</p>
<p>As a result, when subsetting out, say, females, when calling <code>scan1</code>, you only need to subset one of the inputs, and the rest will be automatically subset for you.</p>
<pre><code>out_all &lt;- scan1(probs, pheno, kinship)
out_f &lt;- scan1(probs, pheno[sex==&quot;female&quot;,], kinship)</code></pre>
</div>
<div id="order-of-arguments-when-subsetting-cross-objects" class="section level2">
<h2>Order of arguments when subsetting cross objects</h2>
<p>In R/qtl, when subsetting a cross object, you can use square brackets, like this:</p>
<pre><code>library(qtl)
mycross[chr, ind]</code></pre>
<p>But the order of those two arguments was not very well chosen. It’s better to think of individuals as rows and chromosomes as column, and so put the individuals first.</p>
<p>And so in R/qtl2, we’ve switched the order of arguments: in bracket subsetting of cross objects, individuals now come first.</p>
<pre><code>library(qtl2geno)
mycross[ind, chr]</code></pre>
</div>
<div id="in-scan-of-x-chr-need-to-provide-special-covariates" class="section level2">
<h2>In scan of X chr, need to provide special covariates</h2>
<p>In R/qtl, when you perform a single-QTL scan of the X chromosome, it identifies appropriate covariates to include, to avoid spurious linkage due to sex and cross-direction differences.</p>
<p>In R/qtl2, you need to provide such covariates yourself via the <code>Xcovar</code> argument to <code>scan1()</code>, There is a function in qtl2geno, <code>get_x_covar</code>, for deriving these, but you’re a bit more on your own.</p>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

      </div>
      <hr>
      <footer>
        <p><small>
  <!-- start of Karl's footer; modify this part -->
          <a href="http://creativecommons.org/licenses/by/3.0/"><img src="http://i.creativecommons.org/l/by/3.0/88x31.png" alt="CC BY"/></a> &nbsp;
          <a href="http://kbroman.org">Karl Broman</a>
  <!-- end of Karl's footer; modify this part -->
        </small></p>
      </footer>

    </div>
